goreleaser/goreleaser
# Artifactory: Added basic support to push binaries to an Artifactory store

## Types of changes

- [ ] `make ci` passes on my machine.

### Comments

- `make ci` don't passes on my machine, because of failing `fpm` unit tests. It seems that this is not properly installed on my machine. Will try to make it work.
- Documentation: Only a first draft. Please read through and check if everything is understandable

## Description

This PR adds a new Pipeline to push the built binaries to the artifact story [Artifactory](https://www.jfrog.com/artifactory/) by JFrog.
This feature was requested in #344 and mentioned in #320.

## Notes

* How to push an artifact into Artifactory if described in their documentation: [Artifactory REST API: Example - Deploying an Artifact](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#Artifactory%27sRESTAPI-DeployArtifact)
* As an example on how to build a new Pipeline, [draft: Docker integration #355](https://github.com/goreleaser/goreleaser/pull/355/files) was used
- Supported auth methods against Artifactory, only Basic Authentication with username/password or username/apikey is supported. See docs of Artifactory: [Access Tokens: Basic Authentication](https://www.jfrog.com/confluence/display/RTF/Access+Tokens#AccessTokens-BasicAuthentication)

## .goreleaser.yml

This PR extends the capabilities of ``.
Now it is possible to define multiple Artifactory instances like

```yaml
...
artifactories:
  - target: http://192.168.99.100:8081/artifactory/example-repo-local/goreleaser-tool/{{ .ProjectName }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}
    username: goreleaser
...
```

For every Artifactory, a secret (password or API key) has to be defined as well.
This is done like the `GITHUB_TOKEN` env var.

A new env var is introduced: `ARTIFACTORY_%d_SECRET`.

Means for the first configured artifactory, the password / api key will be stored in `ARTIFACTORY_0_SECRET`. For the second configured instance in `ARTIFACTORY_1_SECRET` and so on.

Both new keys (`target` and `username`) are mandatory if an entry in `artifactories` is set.
Defaults were not set in `pipeline/defaults/defaults.go`, because both keys are mandatory.

## How to test

You can start a Artifactory instance via Docker:

```sh
$ docker run --name artifactory -d -p 8081:8081 docker.bintray.io/jfrog/artifactory-oss:latest
```

If you have issues running this container, please check their documentation about this: [Installing Artifactory > Installing with Docker](https://www.jfrog.com/confluence/display/RTF/Installing+with+Docker).

I assume that your Docker Machine IP is 192.168.99.100 and this will be used in this test example.
Your instance is now available at http://192.168.99.100:8081/ 

To run first a manual test via curl, you can run

```sh
$ echo "This is a file" >> mybin
$ curl -u admin:admin -X PUT "http://192.168.99.100:8081/artifactory/example-repo-local/manual-test/mybin-v1.0.0/mybin.txt" -T ./mybin
```

In this request i assume that you chose `admin` as an username and `admin` as a password during the Artifactory start screen installation. When everything works you should see a response similar to this:

```json
{
  "repo" : "example-repo-local",
  "path" : "/manual-test/mybin-v1.0.0/mybin.txt",
  "created" : "2017-11-28T18:00:45.436Z",
  "createdBy" : "admin",
  "downloadUri" : "http://192.168.99.100:8081/artifactory/manual-test/mybin-v1.0.0/mybin.txt",
  "mimeType" : "application/octet-stream",
  "size" : "9",
  "checksums" : {
    "sha1" : "65d01857a69f14ade727fe1ceee0f52a264b6e57",
    "md5" : "a55e303e7327dc871a8e2a84f30b9983",
    "sha256" : "ead9b172aec5c24ca6c12e85a1e6fc48dd341d8fac38c5ba00a78881eabccf0e"
  },
  "originalChecksums" : {
    "sha256" : "ead9b172aec5c24ca6c12e85a1e6fc48dd341d8fac38c5ba00a78881eabccf0e"
  },
  "uri" : "http://192.168.99.100:8081/artifactory/example-repo-local/manual-test/mybin-v1.0.0/mybin.txt"
}
```

Now feel free to add this piece into a `.goreleaser.yml` of your choice

```yaml
artifactories:
  - target: http://192.168.99.100:8081/artifactory/example-repo-local/goreleaser-tool/{{ .ProjectName }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}
    username: goreleaser
```

and do a release.

When you are browsing into your Artifactory to your artifacts (file browser on the left) and you expand the tree, you will see that all your binaries are stored.

## Implementation comments

- Parts of the code of making the HTTP request + unit tests were inspired by [google/go-github](https://github.com/google/go-github)
- Is it "allowed" to import `"github.com/goreleaser/goreleaser/internal/buildtarget"` inside a pipeline?
- Personally i am not very happy about the solution to store the secrets in a format like `ARTIFACTORY_%d_SECRET`. Do you have a better idea? Maybe a `name` key per artifactory instance and then name the env vars like `ARTIFACTORY_<NAME>_SECRET`?

## Feedback

I would love to get feedback from you about the implementation. Maybe there are helper functions in the goreleaser code itself that we are able to use here.
